batchproducer/batchproducer.go

```diff
@@ -237,9 +257,21 @@ func (b *batchProducer) run() {
 
 // from/for interface Producer
 func (b *batchProducer) Stop() error {
-	if b.isRunning() {
-		b.stop <- true
+	b.runningMu.Lock()
+	defer b.runningMu.Unlock()
+
+	if !b.running {
+		return ErrAlreadyStopped
```

Comment:
Same as above, we can reduce contention here releasing the lock right after the critical section.